Use meme generator if saved message is an URL

xarantolus opened the pull request:
See suggestion 405 on the TUM CS Discord server.

This allows users to caption memes right from the bot commands, e.g. the following will work:

```
$1984 Stop spamming
```

will send the following link (if the message `1984` is set to `https://media.tenor.com/qmSIzc-H7vIAAAAC/1984.gif`):

```
https://api.memegen.link/images/custom/_/Stop_spamming.gif?background=https://media.tenor.com/qmSIzc-H7vIAAAAC/1984.gif
```

Discord will of course embed that link correctly.

I tested this on a test server and it mostly works except for this known problem:
* `/` is not escaped correctly, but it appears that the used API does not accept it correctly even when escaped correctly (even when setting `safe=''` in `quote`)

xarantolus committed:
Use meme generator if saved message is an URL

timschumi commented on cogs/msg.py:
```diff
@@ -74,7 +76,22 @@ async def on_message(self, message):
+            except:
+                return False
+
+        rest = message.clean_content[search.end():].strip()
```
I feel like we should make a specific syntax for this, that can be built into the regex. Or just make it a command, `!msg memegen <shorthand> <text>` instead of weirdly overloading the already implicit behaviour?

xarantolus replied:
I chose `!msg caption` to not collide with the enable/disable commands

timschumi commented on cogs/msg.py:
```diff
@@ -74,7 +76,22 @@ async def on_message(self, message):
+        if isUrl(text) and len(rest) > 0:
+            rest = re.sub(r'\s+', '_', rest)
+            rest = quote(rest, safe='/')
+            text = f"https://api.memegen.link/images/custom/_/{rest}.gif?background={text}"
```
This looks prone to injecting pings and other fun things. However, I'm not sure how to sanitize it best.

xarantolus replied:
I added both an escape using discord.py's `escape_mentions` and also looked into better escaping for the memegen API

timschumi commented on cogs/msg.py:
```diff
@@ -1,4 +1,6 @@
 import re
+from urllib.parse import urlparse, quote
+
```
Stray newline?

timschumi commented on cogs/msg.py:
```diff
@@ -74,7 +76,22 @@ async def on_message(self, message):
+                return False
+
+        rest = message.clean_content[search.end():].strip()
+        if isUrl(text) and len(rest) > 0:
```
This should also be disableable using a config option (or rather, a permission) in my opinion.

xarantolus replied:
Added a permission and per-shorthand enable/disable

timschumi commented on cogs/msg.py:
```diff
@@ -74,7 +76,22 @@ async def on_message(self, message):
-        await message.channel.send(result[0][1])
+        text = result[0][1]
+
+        def isUrl(url):
+            try:
+                result = urlparse(url)
+                return all([result.scheme, result.netloc])
+            except:
+                return False
```
How about a `!msg memegen enable <shorthand>` and `!msg memegen disable <shorthand>` instead of determining this at runtime? We could do the URL check once then, and then just store that in a database. (EDIT: Not specifically as a performance optimization, I'd just like to have the server owner be able to configure what people should be able to embed and what not.)

xarantolus replied:
Makes sense 👍

xarantolus committed:
Memegen: allow enable/disable, explicit caption command

xarantolus commented:
Thank you for the review, I adjusted the code :)

The kind of weird type annotation for the `caption` parameter makes sure that clickable mentions and channel mentions are just passed as text, otherwise it would show user/channel IDs there.

I also added logic to make sure that when setting a new text, `allow_memegen` only stays enabled if the new text is also an URL.

Basically now the following commands work:

```
!msg memegen enable <shorthand>
!msg memegen disable <shorthand>
!msg caption <shorthand> <user-defined caption>
```

xarantolus committed:
Pylint fixes & better error mesage

timschumi commented on cogs/msg.py:
```diff
@@ -37,13 +70,78 @@ async def set(self, ctx, name, *, content):
+
+    @msg.command()
+    @basedbot.has_permissions("msg.caption")
+    async def caption(self, ctx, shorthand, *, caption: commands.clean_content(
+        fix_channel_mentions=True, use_nicknames=True
+    )):
```
I wasn't even aware that this worked, but I'd rather have this in a separate statement instead of bloating up the function definition.

timschumi commented on cogs/msg.py:
```diff
@@ -76,12 +174,14 @@ async def on_message(self, message):
 
         await message.channel.send(result[0][1])
 
-
```
This newline is mandated by some PEP guideline (at least the last time that I ran pylint), as it separates the class from other non-class-related functions.

timschumi commented on cogs/msg.py:
```diff
@@ -37,13 +70,78 @@ async def set(self, ctx, name, *, content):
+            await ctx.send("Shorthand not found.")
+            return
+
+        if not result[0][2]:
+            await ctx.send("Meme generation has not been enabled for this shorthand.")
+            return
```
Could probably just `SELECT` only entries where memegen has been enabled to begin with and adapt the "Shorthand not found" message accordingly ("... or captioning has not been enabled for the chosen shorthand").

timschumi reviewed (changes requested):
A few more nitpicks, after that it's probably "yolo, I'm not at home, lets blind merge this anyways"-time.

timschumi commented on cogs/msg.py:
```diff
@@ -37,13 +70,78 @@ async def set(self, ctx, name, *, content):
+            await ctx.send("Shorthand text must be a valid URL to enable meme generation.")
+            return
+
+        safe_caption = escape_mentions(caption)
```
Isn't this already handled by `clean_content`?

xarantolus replied:
True

xarantolus committed:
Remove unnecessary escape and quote URL

xarantolus commented:
Also now the URL is being quoted correctly, which would have led to problems when an URL with `?` is saved as text

xarantolus committed:
Move caption check to SQL, clean message content
